File: ios/Classes/FlutterTwilioPlugin.swift
import Flutter
import UIKit
import AVFoundation
import PushKit
import TwilioVoice
import CallKit
import UserNotifications

/// Entry point for the `flutter_twilio` iOS plugin.
///
/// Implements the Pigeon-generated [VoiceHostApi] and delegates each typed
/// method to a single-purpose handler. All errors are translated via
/// [FlutterTwilioError] so Dart observers receive the stable codes documented
/// in the spec.
///
/// Asynchronous call-lifecycle events are pushed back to Dart via the
/// [VoiceFlutterApi] installed on the [TVEventEmitter] — never via a
/// [FlutterEventChannel].
///
/// ### Note on `@objc(FlutterTwilioPlugin)`
///
/// Swift normally mangles Objective-C class names with the module prefix
/// (`flutter_twilio.FlutterTwilioPlugin`). The explicit `@objc` annotation
/// pins the runtime name so Flutter's generated `GeneratedPluginRegistrant.m`
/// — which selects plugins by plain Objective-C class name — can resolve us
/// unambiguously, and so that a consumer app with a lingering `twilio_voice`
/// dependency in its tree cannot collide with our registration key (a
/// failure mode the upstream plugin hit as `Duplicate plugin key`).
@objc(FlutterTwilioPlugin)
public class FlutterTwilioPlugin: NSObject, FlutterPlugin, VoiceHostApi {

    // MARK: - Shared state

    let state = TVPluginState()
    let emitter = TVEventEmitter()
    let audioDevice: DefaultAudioDevice = DefaultAudioDevice()

    // MARK: - CallKit / PushKit infrastructure

    let callObserver = CXCallObserver()
    let callKitProvider: CXProvider
    let callKitCallController: CXCallController
    let voipRegistry: PKPushRegistry

    // MARK: - Handlers

    let audioHandler: TVAudioHandler
    let permissionHandler: TVPermissionHandler
    let callHandler: TVCallHandler
    let registrationHandler: TVRegistrationHandler

    // MARK: - Pigeon Flutter API

    private var flutterApi: VoiceFlutterApi?

    static var appName: String {
        (Bundle.main.infoDictionary?["CFBundleName"] as? String) ?? "FlutterTwilio"
    }

    public override init() {
        // PushKit + CallKit configuration.
        voipRegistry = PKPushRegistry(queue: DispatchQueue.main)
        let configuration = CXProviderConfiguration(localizedName: FlutterTwilioPlugin.appName)
        configuration.maximumCallGroups = 1
        configuration.maximumCallsPerCallGroup = 1
        configuration.includesCallsInRecents = true
        callKitProvider = CXProvider(configuration: configuration)
        callKitCallController = CXCallController()

        // Build handlers — plain Swift, no MethodChannel plumbing.
        audioHandler = TVAudioHandler(state: state, emitter: emitter, audioDevice: audioDevice)
        permissionHandler = TVPermissionHandler()
        callHandler = TVCallHandler(
            state: state,
            emitter: emitter,
            audioHandler: audioHandler,
            permissionHandler: permissionHandler,
            callController: callKitCallController,
            callKitProvider: callKitProvider
        )
        registrationHandler = TVRegistrationHandler(state: state, emitter: emitter)

        super.init()

        callObserver.setDelegate(self, queue: DispatchQueue.main)
        callKitProvider.setDelegate(self, queue: nil)
        voipRegistry.delegate = self
        voipRegistry.desiredPushTypes = Set([PKPushType.voIP])
        UNUserNotificationCenter.current().delegate = self

        // The TwilioVoice `Call.Delegate` + `NotificationDelegate` conformances
        // live on the plugin; the call handler grabs us via `delegateOwner` so
        // `ConnectOptions` can point at the same instance.
        callHandler.delegateOwner = self

        // Active-call snapshot provider for incoming events.
        emitter.activeCallProvider = { [weak self] in
            self?.callHandler.getActiveCall()
        }
    }

    deinit {
        callKitProvider.invalidate()
    }

    /// Single-instance guard. Flutter's plugin registrar should only call this
    /// once per engine attach, but hot-restart + embedded-engine scenarios can
    /// trigger a second call; re-entering `VoiceHostApiSetup.setUp` with a
    /// fresh instance would orphan the previous Pigeon handler and leak the
    /// old `CXProvider`. Guard by discarding repeat calls.
    private static var registered = false

    public static func register(with registrar: FlutterPluginRegistrar) {
        guard !registered else {
            NSLog("[flutter_twilio] register(with:) called twice; ignoring the second attach.")
            return
        }
        registered = true

        let instance = FlutterTwilioPlugin()
        VoiceHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
        instance.flutterApi = VoiceFlutterApi(binaryMessenger: registrar.messenger())
        if let api = instance.flutterApi { instance.emitter.attach(api: api) }
        registrar.addApplicationDelegate(instance)
    }

    // MARK: - VoiceHostApi

    func setAccessToken(token: String, completion: @escaping (Result<Void, Error>) -> Void) {
        guardCall(completion) { try self.registrationHandler.setAccessToken(token) }
    }

    func register(completion: @escaping (Result<Void, Error>) -> Void) {
        guardCall(completion) { try self.registrationHandler.register() }
    }

    func unregister(completion: @escaping (Result<Void, Error>) -> Void) {
        guardCall(completion) { try self.registrationHandler.unregister() }
    }

    func place(request: PlaceCallRequest, completion: @escaping (Result<ActiveCallDto, Error>) -> Void) {
        let extra: [String: String] = request.extraParameters?.reduce(
            into: [String: String]()
        ) { acc, pair in
            if let k = pair.key, let v = pair.value { acc[k] = v }
        } ?? [:]
        // place() is async — the continuation is resolved from
        // `callDidStartRinging` (success) / `callDidFailToConnect` (failure).
        Task { [callHandler] in
            do {
                let dto = try await callHandler.place(
                    to: request.to,
                    from: request.from,
                    extra: extra
                )
                completion(.success(dto))
            } catch let pe as PigeonError {
                completion(.failure(pe))
            } catch {
                let ns = error as NSError
                if ns.domain == "com.twilio.voice" || ns.domain.contains("Twilio") {
                    completion(.failure(FlutterTwilioError.fromTwilio(error)))
                } else {
                    completion(.failure(FlutterTwilioError.unknown(error)))
                }
            }
        }
    }

    func answer(completion: @escaping (Result<Void, Error>) -> Void) {
        guardCall(completion) { try self.callHandler.answer() }
    }

    func reject(completion: @escaping (Result<Void, Error>) -> Void) {
        guardCall(completion) { try self.callHandler.reject() }
    }

    func hangUp(completion: @escaping (Result<Void, Error>) -> Void) {
        guardCall(completion) { try self.callHandler.hangUp() }
    }

    func setMuted(muted: Bool, completion: @escaping (Result<Void, Error>) -> Void) {
        guardCall(completion) { try self.callHandler.setMuted(muted) }
    }

    func setOnHold(onHold: Bool, completion: @escaping (Result<Void, Error>) -> Void) {
        guardCall(completion) { try self.callHandler.setOnHold(onHold) }
    }

    func setSpeaker(onSpeaker: Bool, completion: @escaping (Result<Void, Error>) -> Void) {
        guardCall(completion) { try self.audioHandler.setSpeaker(onSpeaker) }
    }

    func sendDigits(digits: String, completion: @escaping (Result<Void, Error>) -> Void) {
        guardCall(completion) { try self.callHandler.sendDigits(digits) }
    }

    func getActiveCall(completion: @escaping (Result<ActiveCallDto?, Error>) -> Void) {
        guardCall(completion) { self.callHandler.getActiveCall() }
    }

    func hasMicPermission(completion: @escaping (Result<Bool, Error>) -> Void) {
        guardCall(completion) { self.permissionHandler.hasMicPermission() }
    }

    func requestMicPermission(completion: @escaping (Result<Bool, Error>) -> Void) {
        permissionHandler.requestMicPermission { granted in
            completion(.success(granted))
        }
    }

    // MARK: - Helpers

    private func guardCall<T>(
        _ completion: @escaping (Result<T, Error>) -> Void,
        _ body: () throws -> T
    ) {
        do {
            completion(.success(try body()))
        } catch let pe as PigeonError {
            completion(.failure(pe))
        } catch {
            let ns = error as NSError
            if ns.domain == "com.twilio.voice" || ns.domain.contains("Twilio") {
                completion(.failure(FlutterTwilioError.fromTwilio(error)))
            } else {
                completion(.failure(FlutterTwilioError.unknown(error)))
            }
        }
    }
}

// MARK: - UIWindow convenience (used by notification presentation code paths)
extension UIWindow {
    func topMostViewController() -> UIViewController? {
        return topViewController(for: rootViewController)
    }

    func topViewController(for rootViewController: UIViewController?) -> UIViewController? {
        guard let rootViewController = rootViewController else { return nil }
        guard let presentedViewController = rootViewController.presentedViewController else { return rootViewController }
        switch presentedViewController {
        case is UINavigationController:
            let nav = presentedViewController as! UINavigationController
            return topViewController(for: nav.viewControllers.last)
        case is UITabBarController:
            let tab = presentedViewController as! UITabBarController
            return topViewController(for: tab.selectedViewController)
        default:
            return topViewController(for: presentedViewController)
        }
    }
}

File: android/src/main/kotlin/com/dev/flutter_twilio/FlutterTwilioPlugin.kt
package com.dev.flutter_twilio

import android.util.Log
import com.dev.flutter_twilio.generated.ActiveCallDto
import com.dev.flutter_twilio.generated.AudioRoute
import com.dev.flutter_twilio.generated.AudioRouteInfo
import com.dev.flutter_twilio.generated.CallDirection
import com.dev.flutter_twilio.generated.FlutterError
import com.dev.flutter_twilio.generated.PlaceCallRequest
import com.dev.flutter_twilio.generated.VoiceConfig
import com.dev.flutter_twilio.generated.VoiceFlutterApi
import com.dev.flutter_twilio.generated.VoiceHostApi
import com.dev.flutter_twilio.audio.TVAudioRouteListener
import com.dev.flutter_twilio.audio.TVAudioRouter
import com.dev.flutter_twilio.generated.CallEventType
import com.dev.flutter_twilio.handler.TVAudioMethodHandler
import com.dev.flutter_twilio.handler.TVCallMethodHandler
import com.dev.flutter_twilio.handler.TVPermissionMethodHandler
import com.dev.flutter_twilio.handler.TVRegistrationMethodHandler
import com.dev.flutter_twilio.service.TVCallManager
import com.dev.flutter_twilio.service.VoiceConfigLocal
import com.dev.flutter_twilio.storage.StorageImpl
import com.dev.flutter_twilio.tone.TVTonePlayer
import com.twilio.voice.CallException
import com.twilio.voice.RegistrationException
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.PluginRegistry.RequestPermissionsResultListener

/**
 * Entry point for the `flutter_twilio` Android plugin.
 *
 * Implements the Pigeon-generated [VoiceHostApi] and delegates each typed
 * method to a single-purpose handler. All errors are translated via
 * [FlutterTwilioError] so Dart observers receive the stable codes documented
 * in the spec.
 *
 * Asynchronous call-lifecycle events are pushed back to Dart via the
 * [VoiceFlutterApi] installed on [TVEventEmitter] — never via an
 * [io.flutter.plugin.common.EventChannel].
 */
class FlutterTwilioPlugin :
    FlutterPlugin,
    ActivityAware,
    RequestPermissionsResultListener,
    VoiceHostApi {

    companion object {
        private const val TAG = "FlutterTwilioPlugin"
    }

    private val state = TVPluginState()
    private val emitter = TVEventEmitter()
    private val callEventsReceiver = TVCallEventsReceiver(state, emitter)

    private lateinit var callHandler: TVCallMethodHandler
    private lateinit var audioHandler: TVAudioMethodHandler
    private lateinit var permissionHandler: TVPermissionMethodHandler
    private lateinit var registrationHandler: TVRegistrationMethodHandler
    private lateinit var audioRouter: TVAudioRouter
    private var audioRouteListener: TVAudioRouteListener? = null

    private var flutterApi: VoiceFlutterApi? = null

    override fun onAttachedToEngine(binding: FlutterPluginBinding) {
        val context = binding.applicationContext
        state.context = context
        state.storage = StorageImpl(context)

        val messenger = binding.binaryMessenger
        VoiceHostApi.setUp(messenger, this)
        flutterApi = VoiceFlutterApi(messenger).also { emitter.attach(it) }

        audioRouter = TVAudioRouter(context)
        callHandler = TVCallMethodHandler(state, emitter)
        audioHandler = TVAudioMethodHandler(state, emitter, audioRouter)
        permissionHandler = TVPermissionMethodHandler(state, emitter)
        registrationHandler = TVRegistrationMethodHandler(state, emitter)

        TVCallManager.init(context)
        TVCallManager.applyConfig(
            VoiceConfigLocal.default,
            TVTonePlayer(context),
            TVTonePlayer(context),
            TVTonePlayer(context),
        )
        TVCallManager.listener = callEventsReceiver

        ActiveCallSnapshotter.provider = { snapshotActiveCall() }

        audioRouteListener = TVAudioRouteListener(context).apply {
            start { route ->
                emitter.emit(CallEventType.AUDIO_ROUTE_CHANGED, audioRoute = route)
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        Log.d(TAG, "onDetachedFromEngine")
        VoiceHostApi.setUp(binding.binaryMessenger, null)
        audioRouteListener?.stop()
        audioRouteListener = null
        emitter.detach()
        flutterApi = null
        TVCallManager.listener = null
        ActiveCallSnapshotter.provider = null
        state.context = null
    }

    // region ActivityAware

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        state.activity = binding.activity
        binding.addRequestPermissionsResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        state.activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        state.activity = binding.activity
        binding.addRequestPermissionsResultListener(this)
    }

    override fun onDetachedFromActivity() {
        state.activity = null
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray,
    ): Boolean = permissionHandler.onPermissionsResult(requestCode, permissions, grantResults)

    // endregion

    // region VoiceHostApi

    override fun setAccessToken(token: String, callback: (Result<Unit>) -> Unit) =
        guard(callback) { registrationHandler.setAccessToken(token) }

    override fun register(callback: (Result<Unit>) -> Unit) =
        guard(callback) { registrationHandler.register() }

    override fun unregister(callback: (Result<Unit>) -> Unit) =
        guard(callback) { registrationHandler.unregister() }

    override fun place(request: PlaceCallRequest, callback: (Result<ActiveCallDto>) -> Unit) =
        guard(callback) {
            val extra: Map<String, String> = request.extraParameters
                ?.filter { (k, v) -> k != null && v != null }
                ?.mapKeys { (k, _) -> k!! }
                ?.mapValues { (_, v) -> v!! }
                ?: emptyMap()
            callHandler.place(to = request.to, from = request.from, extra = extra)
            snapshotActiveCall()
                ?: throw FlutterTwilioError.of(
                    "connection_error",
                    "Outgoing call placed but no active-call snapshot is available",
                )
        }

    override fun answer(callback: (Result<Unit>) -> Unit) =
        guard(callback) { callHandler.answer() }

    override fun reject(callback: (Result<Unit>) -> Unit) =
        guard(callback) { callHandler.reject() }

    override fun hangUp(callback: (Result<Unit>) -> Unit) =
        guard(callback) { callHandler.hangUp() }

    override fun setMuted(muted: Boolean, callback: (Result<Unit>) -> Unit) =
        guard(callback) { callHandler.setMuted(muted) }

    override fun setOnHold(onHold: Boolean, callback: (Result<Unit>) -> Unit) =
        guard(callback) { callHandler.setOnHold(onHold) }

    override fun setSpeaker(onSpeaker: Boolean, callback: (Result<Unit>) -> Unit) =
        guard(callback) { audioHandler.setSpeakerLegacy(onSpeaker) }

    override fun sendDigits(digits: String, callback: (Result<Unit>) -> Unit) =
        guard(callback) { callHandler.sendDigits(digits) }

    override fun getActiveCall(callback: (Result<ActiveCallDto?>) -> Unit) =
        guard(callback) { snapshotActiveCall() }

    override fun hasMicPermission(callback: (Result<Boolean>) -> Unit) =
        guard(callback) { permissionHandler.hasMicPermission() }

    override fun requestMicPermission(callback: (Result<Boolean>) -> Unit) {
        try {
            permissionHandler.requestMicPermission { granted ->
                callback(Result.success(granted))
            }
        } catch (t: Throwable) {
            callback(Result.failure(mapError(t)))
        }
    }

    override fun configure(config: VoiceConfig, callback: (Result<Unit>) -> Unit) =
        guard(callback) { /* TODO(Task 22): apply VoiceConfig */ }

    override fun setAudioRoute(route: AudioRoute, callback: (Result<Unit>) -> Unit) =
        guard(callback) { audioHandler.setAudioRoute(route) }

    override fun getAudioRoute(callback: (Result<AudioRoute>) -> Unit) =
        guard(callback) { audioHandler.getAudioRoute() }

    override fun listAudioRoutes(callback: (Result<List<AudioRouteInfo>>) -> Unit) =
        guard(callback) { audioHandler.listAudioRoutes() }

    override fun bringAppToForeground(callback: (Result<Unit>) -> Unit) =
        guard(callback) {
            throw FlutterTwilioError.of("not_initialized", "bringAppToForeground not yet wired")
        }

    // endregion

    // region Helpers

    private inline fun <T> guard(
        noinline cb: (Result<T>) -> Unit,
        crossinline body: () -> T,
    ) {
        try {
            cb(Result.success(body()))
        } catch (t: Throwable) {
            cb(Result.failure(mapError(t)))
        }
    }

    private fun mapError(t: Throwable): Throwable = when (t) {
        is FlutterError -> t
        is CallException -> FlutterTwilioError.fromTwilio(t)
        is RegistrationException -> FlutterTwilioError.fromRegistration(t)
        else -> FlutterTwilioError.unknown(t)
    }

    /**
     * Build an [ActiveCallDto] from the current [TVCallManager] snapshot,
     * or `null` when no call is active or pending.
     */
    private fun snapshotActiveCall(): ActiveCallDto? {
        if (!TVCallManager.hasActiveCall()) return null
        val sid = TVCallManager.getActiveCallSid() ?: "unknown"
        val direction = when (TVCallManager.callDirection) {
            com.dev.flutter_twilio.types.CallDirection.INCOMING -> CallDirection.INCOMING
            com.dev.flutter_twilio.types.CallDirection.OUTGOING -> CallDirection.OUTGOING
        }
        val custom: Map<String?, String?> = TVCallManager.activeCustomParameters
            .mapKeys { (k, _) -> k as String? }
            .mapValues { (_, v) -> v as String? }
        return ActiveCallDto(
            sid = sid,
            from = TVCallManager.activeCallFrom,
            to = TVCallManager.activeCallTo,
            direction = direction,
            startedAt = TVCallManager.callStartedAtMillis,
            isMuted = state.isMuted,
            isOnHold = state.isHolding,
            isOnSpeaker = state.isSpeakerOn,
            currentRoute = audioRouter.current(),
            connectedAt = TVCallManager.connectedAtMillis,
            customParameters = custom,
        )
    }

    // endregion
}
